common/src/main/java/io/pravega/common/util/btree/BTreeIndex.java

```diff
@@ -790,14 +792,17 @@ private ByteArraySegment generateMinKey() {
         long rootMinOffset = lastPage.getMinOffset();
         assert rootMinOffset >= 0 : "root.MinOffset not set";
         return this.write.apply(pages, oldOffsets, rootMinOffset, timeout)
-                         .thenApply(indexLength -> setState(indexLength, rootOffset, rootLength).length);
+                .thenApply(indexLength -> {
+                    setState(indexLength, rootOffset, rootLength);
```

Comment:
Will some sort of compare-and-set help here in preventing races?
You know the expected previous **state**

Reply:
It's not just in memory state that I care about. What I noticed is that, at least for NFS Tier 2 mounts, even if I send a 1MB data block, if we crash while writing it, part of it may have been durably stored onto whatever is behind the NFS mount. While not a problem for regular stream segments (since they're append-only and we can always resume from where we left off), for BTreeIndex it's different. It needs the file (at least the last 12 bytes of it) to be structured, otherwise it won't be able to read the BTree Pages properly. If we wanted to write 10 pages, for example, but crashed after writing 6.5 of them, then the data structure won't be able to recover afterwards.

Going back to your question. I need to persist **somewhere** the pointer to the root page of the BTree index. Previously it was the last 12 bytes of the Attribute File, but since that's not reliable, one of the changes in this PR is to also store it in the main Segment's Core Attributes (these attributes have a different lifecycle so they're not susceptible to this problem). If we crash in the middle of a write (or just after a good write but before persisting the new pointer), we'll always be able to recover from the previous BTree Index pointer, and try again.

Please wait until I finish this PR. I only submitted it so I can get some Travis builds done and get some system tests executed, but I am not ready for review. I'll add the appropriate description and explanations when I'm done, most likely in a few days.